pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,175 @@
+package distributor
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"github.com/weaveworks/common/httpgrpc"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/weaveworks/common/mtime"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	longPollDuration = 10 * time.Second
+)
+
+// Track the replica we're accepting samples from
+// for each HA cluster we know about.
+type tracker struct {
+	logger log.Logger
+	cfg    HATrackerConfig
+
+	// Replicas we are accepting samples from.
+	electedLock sync.RWMutex
+	elected     map[string]ReplicaDesc
+	done        chan struct{}
+	cancel      context.CancelFunc
+
+	client ring.KVClient
+}
+
+// HATrackerConfig contains the configuration require to
+// create a HA Tracker.
+type HATrackerConfig struct {
+	ReplicaLabel string
+	ClusterLabel string
+
+	// We should only update the timestamp if the difference
+	// between the stored timestamp and the time we received a sample at
+	// is more than this duration.
+	UpdateTimeout time.Duration
+	// We should only failover to accepting samples from a replica
+	// other than the replica written in the KVStore if the difference
+	// between the stored timestamp and the time we received a sample is
+	// more than this duration
+	FailoverTimeout time.Duration
+
+	KVStore ring.KVConfig
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *HATrackerConfig) RegisterFlags(f *flag.FlagSet) {
+	f.DurationVar(&cfg.UpdateTimeout, "ha-tracker.update-timeout", time.Second, "Update the timestamp in the KV store for a given cluster/replica only after this amount of time has passed since the current stored timestamp.")
```

Comment:
This means we're doing 1 QPS per user per distributor.  I'd say we're limited to 100QPS to consul.  Our biggest cluster has ~10 and ~60 active users.  So this is 600QPS.  Might need to tweak this down...